python/private/pypi_index.bzl

```diff
@@ -0,0 +1,302 @@
+    """Read SimpleAPI.
+
+    Args:
+        ctx: The module_ctx or repository_ctx.
+        url: The url parameter that can be passed to ctx.download.
+        envsubst_vars: The env vars to do env sub before downloading.
+        **download_kwargs: Any extra params to ctx.download.
+            Note that output and auth will be passed for you.
+
+    Returns:
+        A similar object to what `download` would return except that in result.out
+        will be the parsed simple api contents.
+    """
+    # TODO: Add a test that env subbed index urls do not leak into the lock file.
+
+    if type(url) == type([]) and len(url) > 1:
+        fail("Only a single url is supported")
+
+    url = url if type(url) == type("") else url[0]
+
+    output_str = url
+    for char in [".", ":", "/", "\\", "$", "[", "]", "{", "}", "'", "\"", "-"]:
+        output_str = output_str.replace(char, "_")
+
+    output = ctx.path(output_str.strip("_").lower() + ".html")
+
+    real_url = envsubst(
+        url,
+        envsubst_vars,
+        ctx.getenv if hasattr(ctx, "getenv") else ctx.os.environ.get,
+    )
+
+    download = ctx.download(
+        url = [real_url],
+        output = output,
+        auth = get_auth(ctx, [real_url]),
+        **download_kwargs
+    )
+
+    return struct(
+        contents = lambda: _read_contents(
+            ctx,
+            download.wait() if download_kwargs.get("block") == False else download,
+            output,
+            url,
+        ),
+    )
+
+def _read_contents(ctx, result, output, url):
+    if not result.success:
+        fail("Failed to download from {}: {}".format(url, result))
+
+    html = ctx.read(output)
+    return get_packages(url, html)
+
+def get_packages_from_requirements(requirements_files):
+    """Get Simple API sources from a list of requirements files and merge them.
+
+    Args:
+        requirements_files(list[str]): A list of requirements files contents.
+
+    Returns:
+        A struct with `simpleapi` attribute that contains a dict of normalized package
+        name to a list of shas that we should index.
+    """
+    want_packages = {}
+    for contents in requirements_files:
+        parse_result = parse_requirements(contents)
+        for distribution, line in parse_result.requirements:
+            distribution = normalize_name(distribution)
+            shas = want_packages.get(distribution)
+            if not shas:
+                shas = sets.make()
+                want_packages[distribution] = shas
+
+            for sha in get_simpleapi_sources(line).shas:
+                sets.insert(shas, sha)
+
+    return struct(
+        simpleapi = want_packages,
+    )
+
+def get_simpleapi_sources(line):
+    """Get PyPI sources from a requirements.txt line.
+
+    We interpret the spec described in
+    https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers
+
+    Args:
+        line(str): The requirements.txt entry.
```

Comment:
This is a slightly different kind of parsing and we need to somehow separate/extract the `sha256` values. I could modify the original parser, but did not want to update that code in case this has bugs. Maybe I could add a note here to refactor later once we stabilize all of this?

Reply:
Ah ok. I think its fine then.